### DISCUSSION TITLE
Only one of multiple Background Tasks is ever being run per request

### DISCUSSION THREAD

**Discussion by EthanML** [fastapi/fastapi#6098] [Questions] [answered]
### First Check

- [X] I added a very descriptive title to this issue.
- [X] I used the GitHub search to find a similar issue and didn't find it.
- [X] I searched the FastAPI documentation, with the integrated search.
- [X] I already searched in Google "How to X in FastAPI" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/samuelcolvin/pydantic).
- [X] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
- [X] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
from fastapi import APIRouter, BackgroundTasks

router = APIRouter()

def test_func(s: str) -> bool:
    print(f"~~~ test_func: {s}")
    return True


@router.get(
    "/test",
)
def get_test(background_tasks: BackgroundTasks) -> dict:
    background_tasks.add_task(test_func, "one") # Only this task's output will be logged
    background_tasks.add_task(test_func, "two")
    background_tasks.add_task(test_func, "three")
    return {"message": "test"}
```

### Description

- I add multiple background tasks when calling my API endpoint
- I expect each added task to be performed
- Instead, only the first task added appears to ever be run, as indicated by only the first log message appearing

### Operating System

Linux

### Operating System Details

Running via Docker on the `python:3.10.2` base image.

### FastAPI Version

0.85.0

### Python Version

3.10.2

### Additional Context

Uvicorn command being used to start the app is:

```
uvicorn --host=0.0.0.0 --port=8002 app.main:app
```

**Comment by moadennagi** [comment]
Could not reproduce this issue, using the following works ok:
```dockerfile
FROM python:3.10.2

ENV PYTHONUNBUFFERED=1

COPY tmp.py app/

WORKDIR app

RUN python -m venv /opt/venv

RUN /opt/venv/bin/pip install --upgrade pip && \
    /opt/venv/bin/pip install fastapi==0.85.0 uvicorn==0.20.0

CMD ["/opt/venv/bin/uvicorn", "tmp:app", "--reload", "--host=0.0.0.0", "--port=8000"]
```
this is tmp.py
```python
from fastapi import FastAPI, APIRouter, BackgroundTasks

router = APIRouter()

def test_func(s: str) -> bool:
    print(f"~~~ test_func: {s}")
    return True


@router.get(
    "/test",
)
def get_test(background_tasks: BackgroundTasks) -> dict:
    background_tasks.add_task(test_func, "one") # Only this task's output will be logged
    background_tasks.add_task(test_func, "two")
    background_tasks.add_task(test_func, "three")
    return {"message": "test"}

app = FastAPI()
app.include_router(router)

```
curl http://0.0.0.0:8000/test gives this

```
INFO:     Will watch for changes in these directories: ['/app']
INFO:     Uvicorn running on http://0.0.0.0:8000 (Press CTRL+C to quit)
INFO:     Started reloader process [1] using StatReload
INFO:     Started server process [8]
INFO:     Waiting for application startup.
INFO:     Application startup complete.
INFO:     172.17.0.1:56302 - "GET /test HTTP/1.1" 200 OK
~~~ test_func: one
~~~ test_func: two
~~~ test_func: three

```

**Comment by Kludex** [comment]
The latest FastAPI shouldn't have this issue. Can you confirm?

**Comment by EthanML** [comment]
> The latest FastAPI shouldn't have this issue. Can you confirm?

Will give this a go and confirm, thanks. Although @moadennagi not seeing my issue above when using the same version is certainly interesting 🤔

**Comment by Kludex** [comment, marked as answer]
Well... @moadennagi did a great job on providing a code snippet that is self-contained, proving that you don't have any issue with the code snippet provided.

Your issue is that you have a larger code source, that uses a middleware based on `BaseHTTPMiddleware` (like `@app.middleware("http")`), and there was a bug with the interaction with `BackgroundTasks`. Once you bump FastAPI to the latest version, you'll not have the issue anymore.

**Comment by EthanML** [comment]
Confirmed that updating fastapi to `0.88.0` sees this issue resolved, thanks all 👍
